src/types/bytearray.rs

```diff
@@ -23,36 +23,48 @@ impl PyByteArray {
 
     /// Creates a new Python `bytearray` object with an `init` closure to write its contents.
     /// Before calling `init` the bytearray is zero-initialised.
-    ///
-    /// Panics if out of memory.
+    /// * If Python raises a MemoryError on the allocation, `new_with` will return
+    ///   it inside `Err`.
+    /// * If `init` returns `Err(e)`, `new_with` will return `Err(e)`.
+    /// * If `init` returns `Ok(())`, `new_with` will return `Ok(&PyByteArray)`.
     ///
     /// # Example
     /// ```
     /// use pyo3::{prelude::*, types::PyByteArray};
-    /// Python::with_gil(|py| {
+    /// Python::with_gil(|py| -> PyResult<()> {
     ///     let py_bytearray = PyByteArray::new_with(py, 10, |bytes: &mut [u8]| {
     ///         bytes.copy_from_slice(b"Hello Rust");
-    ///     });
+    ///         Ok(())
+    ///     })?;
     ///     let bytearray: &[u8] = unsafe { py_bytearray.as_bytes() };
     ///     assert_eq!(bytearray, b"Hello Rust");
+    ///     Ok(())
     /// });
     /// ```
-    pub fn new_with<F>(py: Python, len: usize, init: F) -> &PyByteArray
+    pub fn new_with<F>(py: Python, len: usize, init: F) -> PyResult<&PyByteArray>
     where
-        F: FnOnce(&mut [u8]),
+        F: FnOnce(&mut [u8]) -> PyResult<()>,
     {
         unsafe {
-            let length = len as ffi::Py_ssize_t;
-            let pyptr = ffi::PyByteArray_FromStringAndSize(std::ptr::null(), length);
-            // Iff pyptr is null, py.from_owned_ptr(pyptr) will panic
-            let pybytearray = py.from_owned_ptr(pyptr);
+            let pyptr =
+                ffi::PyByteArray_FromStringAndSize(std::ptr::null(), len as ffi::Py_ssize_t);
+            // Check for an allocation error and return it
+            if pyptr.is_null() {
+                return Err(PyErr::fetch(py));
+            }
             let buffer = ffi::PyByteArray_AsString(pyptr) as *mut u8;
             debug_assert!(!buffer.is_null());
             // Zero-initialise the uninitialised bytearray
             std::ptr::write_bytes(buffer, 0u8, len);
             // (Further) Initialise the bytearray in init
-            init(std::slice::from_raw_parts_mut(buffer, len));
-            pybytearray
+            match init(std::slice::from_raw_parts_mut(buffer, len)) {
```

Comment:
I've just noticed that if `init` panics, then this will leak memory =(

I'm going to add `Py::into_ref` later today, which I think you should be able to use to avoid this case.

(Solution I'm thinking would be that instead of `if pyptr.is_null()`, you can call `Py::from_owned_ptr_or_err` to take ownership of the pointer. And then `Py::into_ref(py)` at the end of the function.)

Reply:
Thanks for finding this issue! What will `Py::into_ref` do? Will I still have to call `ffi::Py_DECREF(pyptr)` on an `Err` in `init` to deallocate it or will this be taken care of as I now have a `&PyBytes` reference from the beginning?

Reply:
You'll have `Py<PyBytes>` from the beginning, and `.into_ref(py)` will convert this into `&PyBytes`.

It'll take care of deallocating automatically 🚀

Reply:
Ah, I was slightly confused by having both `Py::from_owned_ptr_or_err(py, ptr)` and `py.from_owned_ptr_or_err(ptr)` :)

Reply:
The code for the `new_with` function is actually starting to look very clean and concise - I look forward to using `Py<T>::into_ref(py)`

Reply:
FYI I didn't quite get round to writing `Py::into_ref` today - I wanted to get #1063 merged first because that's quite a large refactoring to exactly the same code. I'll ping here as soon as `Py::into_ref` is available!

Reply:
Thanks!

Reply:
See #1098 - hopefully you'll have this API soon!

Reply:
`Py::into_ref` is now available to use 🚀

Reply:
I just hope I didn't mess up merging anything ...

Reply:
I did fail in one way though and had to rewrite some of my changes (I had prototyped everything with a declaration of `Py::into_ref`) - so I might have missed something